File: 0695-max-area-of-island/0695-max-area-of-island.cpp
class Solution {
public:
    int dirs[4][2] = {
        {-1, 0},
        {1, 0},
        {0, 1},
        {0, -1}
    };

    int bfs(const int n, const int m, vector<vector<int>>& grid, vector<vector<bool>>& vis, int i, int j){
        queue<pair<int, int>> q{};
        vis[i][j] = true;
        q.push({i, j});

        int area = 0;

        while(!q.empty()){
            auto [y, x] = q.front(); q.pop();
            ++area;

            for(int i = 0; i < 4; ++i){
                int dy = y + dirs[i][0], dx = x + dirs[i][1];
                if(dy >= n || dy < 0 || dx >= m || dx < 0) continue;
                if(grid[dy][dx] == 0 || vis[dy][dx]) continue;
                vis[dy][dx] = true;
                q.push({dy, dx});
            }
        }
        return area;
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        const int n = grid.size(), m = grid[0].size();
        int ans = 0;
        vector<vector<bool>> vis(n, vector<bool>(m, false));
        for(int i = 0; i < n; ++i){
            for(int j = 0; j < m; ++j){
                if(grid[i][j] == 0 || vis[i][j]) continue;
                ans = max(ans, bfs(n, m, grid, vis, i, j));
            }
        }
        return ans;
    }
};

File: 0695-max-area-of-island/0695-max-area-of-island.erl
-spec max_area_of_island(Grid :: [[integer()]]) -> integer().
max_area_of_island(Grid) ->
    N = length(Grid),
    M = length(hd(Grid)),
    element(2, lists:foldl(
        fun({X, Y}, {Vis, Res}) ->
            case {sets:is_element({X, Y}, Vis), lists:nth(X, lists:nth(Y, Grid))} of
                {false, 1} -> 
                    {NewVis, Area} = dfs(Grid, sets:add_element({X, Y}, Vis), N, M, X, Y),
                    {NewVis, max(Res, Area)};
                _ -> {Vis, Res}
            end        
        end, {sets:new(), 0}, [{X, Y} || Y <- lists:seq(1, N), X <- lists:seq(1, M)]
    )).

dfs(G, Vis, N, M, X, Y) ->
    lists:foldl(
        fun({_X, _Y}, {Vis, Area}) ->
            DX = X + _X,
            DY = Y + _Y,
            case {sets:is_element({DX, DY}, Vis), DX =< 0, DX > M, DY =< 0, DY > N} of
                {false, false, false, false, false} ->
                    case lists:nth(DX, lists:nth(DY, G)) of
                        0 -> {Vis, Area};
                        1 ->
                            {NewVis, Add} = dfs(G, sets:add_element({DX, DY}, Vis), N, M, DX, DY),
                            {NewVis, Area + Add}
                    end;
                _ -> {Vis, Area}
            end
        end, {Vis, 1}, [{1, 0}, {0, 1}, {-1, 0}, {0, -1}]
    ).